libs/sdk-ui-pivot/src/impl/structure/tableDescriptor.ts

```diff
@@ -0,0 +1,352 @@
+// (C) 2021 GoodData Corporation
+
+import {
+    AnyCol,
+    DataColLeaf,
+    isDataColGroup,
+    isDataColLeaf,
+    isSliceCol,
+    isDataColRootGroup,
+    SliceCol,
+    TableColDefs,
+    TableCols,
+    agColId,
+} from "./tableDescriptorTypes";
+import { ColDef, ColGroupDef, Column } from "@ag-grid-community/all-modules";
+import invariant from "ts-invariant";
+import { IAttributeColumnWidthItem, IMeasureColumnWidthItem } from "../../columnWidths";
+import { IAttributeDescriptor, IMeasureDescriptor } from "@gooddata/sdk-backend-spi";
+import { searchForLocatorMatch } from "./colLocatorMatching";
+import { DataViewFacade } from "@gooddata/sdk-ui";
+import { createHeadersAndColDefs } from "./tableDescriptorFactory";
+import { ISortItem } from "@gooddata/sdk-model";
+import { createSortIndicators, SortIndicator } from "./tableDescriptorSorting";
+import { createSortItemForCol } from "./colSortItemFactory";
+
+/**
+ * Table Descriptor is the entry point to all table structure data and metadata. It contains exhaustive information
+ * about all table columns (column descriptors) and their respective ag-grid column definitions.
+ *
+ * On top of this data and metadata, the table descriptor provides convenience and added value methods.
+ *
+ * Column Descriptors vs ag-grid ColDefs | ColGroupDefs
+ * ----------------------------------------------------
+ *
+ * Column Descriptors (shortened to `Col` sakes of brevity) are our implementation-specific descriptors for
+ * the table columns and their grouping into column groups. They contain all the essential GD-specific metadata about
+ * the content of the respective table column (attribute descriptors, headers, measure descriptors) and additional
+ * structural information.
+ *
+ * Column Descriptors are divided into two groups: slicing column descriptors and data column descriptors. The
+ * slicing columns essentially describe the row attribute columns, while the data columns describe columns that may
+ * contains the computed metric values. The data column is a composite. In tables the scope metrics for one or more
+ * attributes, the table columns will be further grouped by the different attribute elements to which the metrics
+ * were scoped to.
+ *
+ * The ag-grid ColDefs and ColGroupDefs are (naturally) used to construct the ag-grid table :) They are built to
+ * reflect the column descriptors but only contain the information needed by ag-grid itself.
+ *
+ * The important thing to remember is that ColDefs and ColGroupDefs have same colId/groupId as their column descriptor
+ * counterparts.
+ *
+ *
+ * @alpha
+ */
+export class TableDescriptor {
+    public readonly zippedSliceCols: Array<[SliceCol, ColDef]> = [];
+    public readonly zippedLeaves: Array<[DataColLeaf, ColDef]> = [];
```

Comment:
Can you please add some docs for the zipped collections?

Reply:
done